packages/zoe/src/contracts/barterExchange.js

```diff
@@ -0,0 +1,188 @@
+// @ts-check
+
+import harden from '@agoric/harden';
+import { makeZoeHelpers, defaultAcceptanceMsg } from '../contractSupport';
+
+/** @typedef {import('../zoe').ContractFacet} ContractFacet */
+
+/**
+ * The Barter Exchange ignores the keywords in offers. It takes advantage of
+ * Zoe facet parameters by only paying attention the issuers in proposals.
+ *
+ * The want and give amounts are both treated as minimums. Each successful
+ * trader gets their `want` and may trade with counter-parties who specify any
+ * amount up to their specified `give`.
+ */
+export const makeContract = harden(
+  /** @param {ContractFacet} zcf */ zcf => {
+    // bookOrders is a Map of Maps. The first key is the brand of each offer's
+    // give, and the second key is the brand of their want.
+    // for each offer, we store (see extractOfferDetails) the handle, as well as
+    // keywords, brands, and amount for both `give` and `want`. The keywords are
+    // only used to produce the payout.
+    const bookOrders = new Map();
+    let brandToAmountMath;
+
+    const { rejectOffer, getKeys, assertKeywords } = makeZoeHelpers(zcf);
+
+    assertKeywords(harden(['Asset', 'Price']));
+
+    function lookupBookOrders(brandIn, brandOut) {
+      let ordersMap = bookOrders.get(brandIn);
+      if (!ordersMap) {
+        ordersMap = new Map();
+        bookOrders.set(brandIn, ordersMap);
+      }
+      let ordersArray = ordersMap.get(brandOut);
+      if (!ordersArray) {
+        ordersArray = [];
+        ordersMap.set(brandOut, ordersArray);
+      }
+      return ordersArray;
+    }
+
+    const getKeywordAndBrand = (offerHandle, amountKeywordRecord) => {
+      const keywords = getKeys(amountKeywordRecord);
+      if (keywords.length !== 1) {
+        rejectOffer(
+          offerHandle,
+          `A swap requires giving one type of token for another, ${keywords.length} tokens were provided.`,
+        );
+      }
+      return harden({
+        keyword: keywords[0],
+        brand: Object.values(amountKeywordRecord)[0].brand,
+      });
+    };
+
+    function extractOfferDetails(offerHandle, proposal) {
+      const keywordAndBrandIn = getKeywordAndBrand(offerHandle, proposal.give);
+      const { brand: brandIn, keyword: keywordIn } = keywordAndBrandIn;
+      const keywordAndBrandOut = getKeywordAndBrand(offerHandle, proposal.want);
+      const { brand: brandOut, keyword: keywordOut } = keywordAndBrandOut;
+      const {
+        proposal: {
+          give: { [keywordIn]: amountIn },
+          want: { [keywordOut]: amountOut },
+        },
+      } = zcf.getOffer(offerHandle);
+      return {
+        offerHandle,
+        keywordOut,
+        brandOut,
+        amountOut,
+        keywordIn,
+        brandIn,
+        amountIn,
+      };
+    }
+
+    function findMatchingTrade(newDetails, orders) {
+      return orders.find(order => {
+        const { amountIn, amountOut } = order;
+        // see if canTradeWith() would work as well
```

Comment:
I think we can't say that we have "facet parameters" unless our helpers support it too. I think we would need to alter `canTradeWith`.

Reply:
I'm ok with that being a separate PR, but I just want to make sure that we aren't "releasing" facet parameters until then.

Reply:
I can do this with a variant `canTradeWithIgnoreKeywords()`, but I'm not convinced it's worthwhile. The answer tells us that there's a way to allocate the incoming goods from each side among the 'want's on the other, but not how to do so. 

I think barter is better off comparing manually. I'll include the version with the helper for you to look at, but my instinct is to revert to the code in this version of the PR.

I think there are too many ways for offers to produce a false positive response from canTradeWithIgnoreKeywords, so contracts really have to compare the impact with an actual proposed reallocation.